File: Exámenes/2019 Febrero/Haskell 2019/DictionaryWeightedGraph.hs
----------------------------------------------
-- Estructuras de Datos.  2018/19
-- 2º Curso del Grado en Ingeniería [Informática | del Software | de Computadores].
-- Escuela Técnica Superior de Ingeniería en Informática. UMA
--
-- Examen 4 de febrero de 2019
--
-- ALUMNO/NAME:
-- GRADO/STUDIES:
-- NÚM. MÁQUINA/MACHINE NUMBER:
--
-- Weighted Graph implemented by using a dictionary from
-- sources to another dictionary from destinations to weights
----------------------------------------------

module DictionaryWeightedGraph
  ( WeightedGraph
  , WeightedEdge(WE)
  , empty
  , isEmpty
  , mkWeightedGraphEdges
  , addVertex
  , addEdge
  , vertices
  , numVertices
  , edges
  , numEdges
  , successors
  ) where

import Data.List(nub, intercalate)

import qualified DataStructures.Dictionary.AVLDictionary as D

data WeightedEdge a w  = WE a w a deriving Show

instance (Eq a, Eq w) => Eq (WeightedEdge a w) where
  WE u w v == WE u' w' v' = (u==u' && v==v' || u==v' && v==u')
                              && w == w'

instance (Eq a, Ord w) => Ord (WeightedEdge a w) where
  compare (WE _ w _) (WE _ w' _) = compare w w'

data WeightedGraph a w  = WG (D.Dictionary a (D.Dictionary a w))

empty :: WeightedGraph a w
empty = WG D.empty

addVertex :: (Ord a) => WeightedGraph a w -> a -> WeightedGraph a w
addVertex (WG dic) x = WG (D.insert x D.empty dic)

addEdge :: (Ord a, Show a) => WeightedGraph a w -> a -> a -> w -> WeightedGraph a w
addEdge (WG dic) x y p
  | not(D.isDefinedAt x dic || D.isDefinedAt y dic) = error "alguno de los vértices no está definido en el grafo"
  | otherwise = WG (D.insert x y' dic)
    where 
      y' = D.insert y p dic'
      Just dic' = D.valueOf x dic

edges :: (Eq a, Eq w) => WeightedGraph a w -> [WeightedEdge a w]
edges (WG dic) = nub[(WE x p y) | (x, vs) <- D.keysValues dic, (y, p) <- D.keysValues vs]

successors :: (Ord a, Show a) => WeightedGraph a w -> a -> [(a,w)]
successors (WG dic) x 
  |not(D.isDefinedAt x dic) = error "El vértice no está definido en el grafo"
  |otherwise = [(y,p) | (y, p) <- D.keysValues d']
    where
      Just d' = D.valueOf x dic 


-- NO EDITAR A PARTIR DE AQUÍ    
-- DON'T EDIT ANYTHING BELOW THIS COMMENT

vertices :: WeightedGraph a w -> [a]
vertices (WG d) = D.keys d

isEmpty :: WeightedGraph a w -> Bool
isEmpty (WG d) = D.isEmpty d

mkWeightedGraphEdges :: (Ord a, Show a) => [a] -> [WeightedEdge a w] -> WeightedGraph a w
mkWeightedGraphEdges vs es = wg'
  where
    wg = foldl addVertex empty vs
    wg' = foldr (\(WE u w v) wg -> addEdge wg u v w) wg es

numVertices :: WeightedGraph a w -> Int
numVertices = length . vertices

numEdges :: (Eq a, Eq w) => WeightedGraph a w -> Int
numEdges = length . edges

instance (Eq a, Show a, Eq w, Show w) => Show (WeightedGraph a w) where
  show wg  = "DictionaryWeightedGraph("++vs++", "++as++")"
   where
    vs  = "("++ intercalate ", " (map show (vertices wg)) ++")"
    as  = "(" ++ intercalate ", " (map showEdge (edges wg)) ++ ")"
    showEdge (WE x w y)  = intercalate "-" [ show x, show w, show y ]

File: Exámenes/2019 Febrero/Java 2019/javilla/src/dataStructures/graph/DictionaryWeightedGraph.java
/**----------------------------------------------
 * -- Estructuras de Datos.  2018/19
 * -- 2º Curso del Grado en Ingeniería [Informática | del Software | de Computadores].
 * -- Escuela Técnica Superior de Ingeniería en Informática. UMA
 * --
 * -- Examen 4 de febrero de 2019
 * --
 * -- ALUMNO/NAME:
 * -- GRADO/STUDIES:
 * -- NÚM. MÁQUINA/MACHINE NUMBER:
 * --
 * ----------------------------------------------
 */

package dataStructures.graph;

import java.util.Iterator;

import dataStructures.dictionary.Dictionary;
import dataStructures.dictionary.HashDictionary;

import dataStructures.set.Set;
import dataStructures.set.HashSet;
import dataStructures.tuple.Tuple2;

public class DictionaryWeightedGraph<V, W extends Comparable<? super W>> implements WeightedGraph<V, W> {

    static class WE<V1, W1 extends Comparable<? super W1>> implements WeightedEdge<V1, W1> {

		V1 src, dst;
        W1 wght;

        WE(V1 s, V1 d, W1 w) {
            src = s;
            dst = d;
            wght = w;
        }

        public V1 source() {
            return src;
        }

        public V1 destination() {
            return dst;
        }

        public W1 weight() {
            return wght;
        }

        public String toString() {
            return "WE(" + src + "," + dst + "," + wght + ")";
        }

		public int hashCode() {
			// COMPLETAR 
			int code = 17;
            int a = 31;
            code = code*a + wght.hashCode();
            code = code*a + src.hashCode() + dst.hashCode();
			return code;
		}

		public boolean equals(Object obj) {
			
			// COMPLETAR 
			if(! (obj instanceof WeightedGraph.WeightedEdge<?,?>)){
                return false;
            }
            WeightedEdge<V1, W1> we = (WeightedEdge<V1, W1>) obj;
            return (src.equals(we.source()) && dst.equals(we.destination()) || src.equals(we.destination()) && dst.equals(we.source())) && wght.equals(we.weight());
		}
		
		public int compareTo(WeightedEdge<V1, W1> o) {

			// COMPLETAR
			
			return wght.compareTo(o.weight());
		}
    }

    /**
     * Each vertex is associated to a dictionary containing associations
     * from each successor to its weight
     */
    protected Dictionary<V, Dictionary<V, W>> graph;

    public DictionaryWeightedGraph() {
        graph = new HashDictionary<>();
    }


    public void addVertex(V v) {
    	// COMPLETAR
        if(!graph.isDefinedAt(v)){
            graph.insert(v, new HashDictionary<>());
        }
    }

    public void addEdge(V src, V dst, W w) {
    	// COMPLETAR
        if(!graph.isDefinedAt(src) || !graph.isDefinedAt(dst)){
            throw new GraphException("Alguno de los vértices no se encuentra definido");
        } else {
            graph.valueOf(src).insert(dst, w);
        }
    }

    public Set<Tuple2<V, W>> successors(V v) {
    	// COMPLETAR
        if(!graph.isDefinedAt(v)){
            throw new GraphException("El vértice no está definido en el grafo");
        }else {
            Dictionary<V, W> dic = graph.valueOf(v);
            Iterable<Tuple2<V, W>> it = dic.keysValues();
            Set<Tuple2<V, W>> sucs = new HashSet<>();

            for (Tuple2<V, W> suc : it) {
                sucs.insert(suc);
            }
            return sucs;
        }

    }


    public Set<WeightedEdge<V, W>> edges() {
    	// COMPLETAR
    	Set<WeightedEdge<V, W>> edges = new HashSet<>();
        for(V vert : graph.keys()){
            for(V v2 : graph.valueOf(vert).keys()){
                WeightedEdge<V, W> we = new WE<>(vert, v2, graph.valueOf(vert).valueOf(v2));
                edges.insert(we);
            }
        }
        return edges;
    }






    /** DON'T EDIT ANYTHING BELOW THIS COMMENT **/


    public Set<V> vertices() {
        Set<V> vs = new HashSet<>();
        for (V v : graph.keys())
            vs.insert(v);
        return vs;
    }


    public boolean isEmpty() {
        return graph.isEmpty();
    }

    public int numVertices() {
        return graph.size();
    }


    public int numEdges() {
        int num = 0;
        for (Dictionary<V, W> d : graph.values())
            num += d.size();
        return num;
    }


    public String toString() {
        String className = getClass().getSimpleName();
        String s = className + "(vertices=(";

        Iterator<V> it1 = vertices().iterator();
        while (it1.hasNext())
            s += it1.next() + (it1.hasNext() ? ", " : "");
        s += ")";

        s += ", edges=(";
        Iterator<WeightedEdge<V, W>> it2 = edges().iterator();
        while (it2.hasNext())
            s += it2.next() + (it2.hasNext() ? ", " : "");
        s += "))";

        return s;
    }
}
